File: packages/sdk/dart/packages/admin/lib/src/analytics_client.dart
import 'package:edgebase_core/src/generated/api_core.dart';
import 'package:edgebase_core/src/generated/client_wrappers.dart';
import 'generated/admin_api_core.dart';

typedef AnalyticsProperties = Map<String, dynamic>;

class AnalyticsEvent {
  final String name;
  final AnalyticsProperties? properties;
  final int? timestamp;
  final String? userId;

  const AnalyticsEvent({
    required this.name,
    this.properties,
    this.timestamp,
    this.userId,
  });
}

class AnalyticsClient {
  final GeneratedAnalyticsMethods _methods;
  final GeneratedAdminApi _adminCore;

  AnalyticsClient(GeneratedDbApi core, this._adminCore) : _methods = GeneratedAnalyticsMethods(core);

  Future<Map<String, dynamic>> overview([Map<String, String>? options]) async {
    final raw = await _adminCore.queryAnalytics(_buildQuery('overview', options));
    return _asMap(raw);
  }

  Future<List<Map<String, dynamic>>> timeSeries([Map<String, String>? options]) async {
    final raw = await _adminCore.queryAnalytics(_buildQuery('timeSeries', options));
    return _extractList(_asMap(raw)['timeSeries']);
  }

  Future<List<Map<String, dynamic>>> breakdown([Map<String, String>? options]) async {
    final raw = await _adminCore.queryAnalytics(_buildQuery('breakdown', options));
    return _extractList(_asMap(raw)['breakdown']);
  }

  Future<List<Map<String, dynamic>>> topEndpoints([Map<String, String>? options]) async {
    final raw = await _adminCore.queryAnalytics(_buildQuery('topEndpoints', options));
    return _extractList(_asMap(raw)['topItems']);
  }

  Future<void> track(
    String name, [
    AnalyticsProperties? properties,
    String? userId,
  ]) async {
    await trackBatch([
      AnalyticsEvent(name: name, properties: properties, userId: userId),
    ]);
  }

  Future<void> trackBatch(List<AnalyticsEvent> events) async {
    if (events.isEmpty) return;
    await _methods.track({
      'events': events
          .map((event) => {
                'name': event.name,
                if (event.properties != null) 'properties': event.properties,
                if (event.userId != null) 'userId': event.userId,
                'timestamp': event.timestamp ?? DateTime.now().millisecondsSinceEpoch,
              })
          .toList(growable: false),
    });
  }

  Future<dynamic> queryEvents([Map<String, String>? options]) {
    return _adminCore.queryCustomEvents(options);
  }

  Map<String, String> _buildQuery(String metric, Map<String, String>? options) {
    return <String, String>{
      'metric': metric,
      ...?options,
    };
  }
}

Map<String, dynamic> _asMap(dynamic value) {
  if (value is Map<String, dynamic>) return value;
  if (value is Map) {
    return value.map((key, entry) => MapEntry(key.toString(), entry));
  }
  return <String, dynamic>{};
}

List<Map<String, dynamic>> _extractList(dynamic value) {
  if (value is! List) return const [];
  return value
      .whereType<Map>()
      .map((item) => item.map((key, entry) => MapEntry(key.toString(), entry)))
      .toList(growable: false);
}

File: packages/sdk/csharp/packages/admin/AnalyticsClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using EdgeBase.Generated;

namespace EdgeBase.Admin
{
    public sealed class AnalyticsEvent
    {
        public string Name { get; set; } = "";
        public Dictionary<string, object?>? Properties { get; set; }
        public long? Timestamp { get; set; }
        public string? UserId { get; set; }
    }

    public sealed class AnalyticsClient
    {
        private readonly GeneratedAnalyticsMethods _methods;
        private readonly GeneratedAdminApi _adminCore;

        internal AnalyticsClient(GeneratedDbApi core, GeneratedAdminApi adminCore)
        {
            _methods = new GeneratedAnalyticsMethods(core);
            _adminCore = adminCore;
        }

        public Task<Dictionary<string, object?>> OverviewAsync(
            Dictionary<string, string>? options = null,
            CancellationToken ct = default) =>
            _adminCore.QueryAnalyticsAsync(BuildQuery("overview", options), ct);

        public async Task<List<Dictionary<string, object?>>> TimeSeriesAsync(
            Dictionary<string, string>? options = null,
            CancellationToken ct = default)
        {
            var result = await _adminCore.QueryAnalyticsAsync(BuildQuery("timeSeries", options), ct);
            return JsonHelper.ExtractList(result, "timeSeries");
        }

        public async Task<List<Dictionary<string, object?>>> BreakdownAsync(
            Dictionary<string, string>? options = null,
            CancellationToken ct = default)
        {
            var result = await _adminCore.QueryAnalyticsAsync(BuildQuery("breakdown", options), ct);
            return JsonHelper.ExtractList(result, "breakdown");
        }

        public async Task<List<Dictionary<string, object?>>> TopEndpointsAsync(
            Dictionary<string, string>? options = null,
            CancellationToken ct = default)
        {
            var result = await _adminCore.QueryAnalyticsAsync(BuildQuery("topEndpoints", options), ct);
            return JsonHelper.ExtractList(result, "topItems");
        }

        public Task TrackAsync(
            string name,
            Dictionary<string, object?>? properties = null,
            string? userId = null,
            CancellationToken ct = default) =>
            TrackBatchAsync(new[]
            {
                new AnalyticsEvent { Name = name, Properties = properties, UserId = userId }
            }, ct);

        public async Task TrackBatchAsync(
            IEnumerable<AnalyticsEvent> events,
            CancellationToken ct = default)
        {
            var materialized = events?.ToList() ?? new List<AnalyticsEvent>();
            if (materialized.Count == 0)
            {
                return;
            }

            var payload = new
            {
                events = materialized.Select(evt =>
                {
                    var entry = new Dictionary<string, object?>
                    {
                        ["name"] = evt.Name,
                        ["timestamp"] = evt.Timestamp ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                    };
                    if (evt.Properties is { Count: > 0 })
                    {
                        entry["properties"] = evt.Properties;
                    }
                    if (!string.IsNullOrWhiteSpace(evt.UserId))
                    {
                        entry["userId"] = evt.UserId;
                    }
                    return entry;
                }).ToList()
            };

            await _methods.TrackAsync(payload, ct);
        }

        public Task<Dictionary<string, object?>> QueryEventsAsync(
            Dictionary<string, string>? options = null,
            CancellationToken ct = default) =>
            _adminCore.QueryCustomEventsAsync(options ?? new Dictionary<string, string>(), ct);

        private static Dictionary<string, string> BuildQuery(
            string metric,
            Dictionary<string, string>? options)
        {
            var query = new Dictionary<string, string> { ["metric"] = metric };
            if (options == null)
            {
                return query;
            }

            foreach (var item in options)
            {
                query[item.Key] = item.Value;
            }

            return query;
        }
    }
}
